src/monodroid/jni/monodroid-glue.c

```diff
@@ -1278,91 +1278,249 @@ gc_is_bridge_object (MonoObject *object)
+static jobject   jni_arraylist, jni_gcuserpeer;
+static jmethodID jni_arraylist_constructor, jni_arraylist_get, jni_arraylist_add, jni_gcuserpeer_constructor;
+
+// Given a target, extract the bridge_info (if a mono object) and handle. Return success.
+static mono_bool
+load_reference_target (AddReferenceTarget target, MonoJavaGCBridgeInfo** bridge_info, jobject *handle)
+{
+	if (target.is_mono_object) {
+		*bridge_info = get_gc_bridge_info_for_object (target.obj);
+		if (!*bridge_info)
+			return FALSE;
+		mono.mono_field_get_value (target.obj, (*bridge_info)->handle, handle);
+	} else {
+		*handle = target.jobj;
+	}
+	return TRUE;
+}
+
+#if DEBUG
+// Allocate and return a string describing a target
+static char *
+describe_target (AddReferenceTarget target)
+{
+	if (target.is_mono_object) {
+		MonoClass *klass = mono.mono_object_get_class (target.obj);
+		return monodroid_strdup_printf ("object of class %s.%s",
+			mono.mono_class_get_namespace (klass),
+			mono.mono_class_get_name (klass));
+	}
+	else
+		return monodroid_strdup_printf ("<temporary object %p>", target.jobj);
+}
+#endif
+
+// Add a reference from one target to another. If the "from" target is a mono_object, it must be a user peer
+static mono_bool
+add_reference (JNIEnv *env, AddReferenceTarget target, AddReferenceTarget reffed_target)
+{
+	MonoJavaGCBridgeInfo *bridge_info = NULL, *reffed_bridge_info = NULL;
+	jobject handle, reffed_handle;
+
+	if (!load_reference_target (target, &bridge_info, &handle))
+		return FALSE;
+
+	if (!load_reference_target (reffed_target, &reffed_bridge_info, &reffed_handle))
+		return FALSE;
+
+	mono_bool success = add_reference_jobject (env, handle, reffed_handle);
+
+	// Flag MonoObjects so they can be cleared in gc_cleanup_after_java_collection.
+	// Java temporaries do not need this because the entire GCUserPeer is discarded.
+	if (success && target.is_mono_object) {
+		int ref_val = 1;
+		mono.mono_field_set_value (target.obj, bridge_info->refs_added, &ref_val);
+	}
+
+#if DEBUG
+	if (gc_spew_enabled) {
+		char *description = describe_target (target),
+			 *reffed_description = describe_target (reffed_target);
+
+		if (success)
+			log_warn (LOG_GC, "Added reference for %s to %s", description, reffed_description);
+		else
+			log_error (LOG_GC, "Missing monodroidAddReference method for %s", description);
+
+		free (description);
+		free (reffed_description);
+	}
+#endif
+
+	return success;
+}
+
+// Create a target
+static AddReferenceTarget
+target_from_mono_object (MonoObject *obj)
+{
+	AddReferenceTarget result;
+	result.is_mono_object = TRUE;
+	result.obj = obj;
+	return result;
+}
+
+// Create a target
+static AddReferenceTarget
+target_from_jobject (jobject jobj)
+{
+	AddReferenceTarget result;
+	result.is_mono_object = FALSE;
+	result.jobj = jobj;
+	return result;
+}
+
+// See note in gc_prepare_for_java_collection
+static int
+scc_get_stashed_index (MonoGCBridgeSCC *scc)
+{
+	assert (scc->num_objs < 0);
```

Comment:
We want `num_objs` to be *negative*? That doesn't seem right...

Additionally, we should strive to have better asserts. (Yes, I'm not doing good here, glass houses and all that...)

Specifically, we should have a useful error message in here:

    assert ((scc->num_objs < 0) || !"Description for why we're asserting");